intellij/src/saros/intellij/ui/tree/ContactPopMenu.java

```diff
@@ -26,33 +29,67 @@
 import saros.repackaged.picocontainer.annotations.Inject;
 
 /**
- * Contact pop-up menu for selecting a project to share. Opened when right-clicking on a contact.
+ * Contact pop-up menu for selecting a project and module to share. Opened when right-clicking on a
+ * contact.
  */
 class ContactPopMenu extends JPopupMenu {
 
   private static final Logger LOG = Logger.getLogger(ContactPopMenu.class);
 
   @Inject private IWorkspace workspace;
 
-  @Inject private Project project;
-
   private final ContactTreeRootNode.ContactInfo contactInfo;
 
   ContactPopMenu(ContactTreeRootNode.ContactInfo contactInfo) {
     this.contactInfo = contactInfo;
 
-    if (workspace == null || project == null) {
-      SarosPluginContext.initComponent(this);
+    SarosPluginContext.initComponent(this);
+
+    JMenu menuShareProject = new JMenu("Work together on...");
+    menuShareProject.setIcon(IconManager.SESSIONS_ICON);
+
+    List<String> nonCompliantModules = new ArrayList<>();
 
-      if (workspace == null || project == null) {
-        LOG.error("PicoContainer injection failed. Objects still not present after injection.");
+    for (Project project : ProjectManager.getInstance().getOpenProjects()) {
+      Pair<JMenu, List<String>> results = prepareProjectMenu(project);
 
+      if (results == null) {
         return;
       }
+
+      JMenu projectMenu = results.first;
+      List<String> projectNonCompliantModules = results.second;
+
+      menuShareProject.add(projectMenu);
+
+      if (!projectNonCompliantModules.isEmpty()) {
+        nonCompliantModules.addAll(projectNonCompliantModules);
+      }
     }
 
-    JMenu menuShareProject = new JMenu("Work together on...");
-    menuShareProject.setIcon(IconManager.SESSIONS_ICON);
+    if (!nonCompliantModules.isEmpty()) {
+      NotificationPanel.showWarning(
+          MessageFormat.format(
+              Messages.Contact_saros_message_conditional,
+              MessageFormat.format(
+                  Messages.ContactPopMenu_invalid_module_message_condition, nonCompliantModules)),
+          Messages.ContactPopMenu_invalid_module_title);
+    }
+
+    add(menuShareProject);
+  }
+
+  /**
+   * Prepares the JMenu for the given project. The menu contains all sharable modules contained in
+   * the project.
+   *
+   * @param project the project whose menu to prepare
+   * @return a <code>Pair</code> containing the prepared JMenu (first value) and a list of
+   *     incompatible modules (second value) or <code>null</code> if the ModuleManager could not be
+   *     instantiated
+   */
+  private Pair<JMenu, List<String>> prepareProjectMenu(@NotNull Project project) {
```

Comment:
You don't have to do this now: Restructuring the method could improve the readability. The "happy" path is hard to find. I would recommend to split the method into multiple smaller methods.

Reply:
See my reply below; will be done in a separate PR.